### ISSUE TITLE
[BUG] Non-Deterministic Model Responses when the Input Prompt Order Changes

### ISSUE THREAD

**Issue by zcakzhuu** [microsoft/DeepSpeed#6612]
**Describe the bug**
When using deepspeed to run inference using a llama2 model, I see different model responses when the order of the prompts passing into the model changes. But if the order of prompts remain the same, the responses are the same. Is it an expected behaviour? I don't tend to expect the order to change model responses. Please let me know if I am wrong.

**To Reproduce**
1. Simple inference script to reproduce

```
import os
import torch
import deepspeed
import time
from tqdm import tqdm
import jsonlines

from transformers import AutoTokenizer, AutoModelForCausalLM, set_seed

set_seed(42)

local_rank = int(os.getenv("LOCAL_RANK", "0"))
world_size = int(os.getenv("WORLD_SIZE", "1"))

hf_token = '<replace with your huggingface token>'
max_tokens = 4096
batch_size = 1

input_list = [
  "Can you tell me about the important people in your life?",
  "What have been some of the happiest moments in your life? The saddest?",
  "Who has been the biggest influence on your life? What lessons did that person or those people teach you?",
  "Can you tell me about a moment when a person’s kindness made a difference in your life?",
  "What are the most important lessons you’ve learned in life?",
  "What is your earliest memory?"
]

# input_list = [
#   "Who has been the biggest influence on your life? What lessons did that person or those people teach you?",
#   "Can you tell me about the important people in your life?",
#   "Can you tell me about a moment when a person’s kindness made a difference in your life?",
#   "What have been some of the happiest moments in your life? The saddest?",
#   "What is your earliest memory?",
#   "What are the most important lessons you’ve learned in life?",
# ]

input_list = [input_list[x:x+batch_size] for x in range(0, len(input_list), batch_size)]

tokenizer = AutoTokenizer.from_pretrained(
    "meta-llama/Llama-2-7b-chat-hf",
    use_fast=True,
    token=hf_token,
)
model = AutoModelForCausalLM.from_pretrained(
    "meta-llama/Llama-2-7b-chat-hf",
    torch_dtype=torch.float16,
    device_map="cpu",
    token=hf_token,
)

tokenizer.add_special_tokens({"pad_token": '[PAD]'})
model.resize_token_embeddings(len(tokenizer))
model.config.pad_token_id = tokenizer.pad_token_id
model.eval()

model = deepspeed.init_inference(
    model=model,
    mp_size=world_size,
    dtype=torch.float16,
    replace_with_kernel_inject=False,
    max_tokens=max_tokens,
)

output_list = []
for inputs in tqdm(input_list):
    torch.cuda.synchronize()
    
    input_tokens = tokenizer.batch_encode_plus(
        [f"[INST]{inp}[/INST]" for inp in inputs], 
        return_tensors="pt",
        padding="longest",
        truncation=True,
        max_length=4096
        )
    
    for t in input_tokens:
        if torch.is_tensor(input_tokens[t]):
            input_tokens[t] = input_tokens[t].to(local_rank)
    
    outputs = model.generate(**input_tokens, max_length=4096, do_sample=True)
    
    response_outputs = []
    for i, o in zip(input_tokens["input_ids"], outputs):
        o = o[len(i):]
        response_outputs.append(o)
        
    response_outputs = tokenizer.batch_decode(
        response_outputs,
        skip_special_tokens=True,
        clean_up_tokenization_spaces=True
    )
    
    response_outputs = [i.strip() for i in response_outputs]
    
    output_list += response_outputs
    
    torch.cuda.synchronize()
    
    torch.cuda.empty_cache()
    
    if local_rank==0:
        print(response_outputs)
    

#save
if local_rank==0:
    disk_output = []
    for q, o in zip(input_list, output_list):
        disk_output.append({"prompt": q, "output": o})
    
    with jsonlines.open('./run3.jsonl', 'w') as writer:
        writer.write_all(disk_output)
```

3. What packages are required and their versions
jsonlines==3.1.0
torch==2.2.2+cu121
deepspeed==0.12.6
accelerate==0.30.1
Cuda driver 12.1

4. How to run the script
`deepspeed --num_gpus 2 script.py`

I run this on 2xT4s.

As you can see, there are 2 `input_list` in the script. I run the script for each `input_list`. The only thing that changed is the order of prompts passing to the model for response generation. I see some slight response differences between the 2 runs. However, if I keep order constant and run for the same `input_list` twice, the responses are exactly the same. 

**Expected behavior**
I would expect the order to not impact the responses. The responses should be somewhat deterministic as set_seed is set. 

**ds_report output**
```
[2024-10-08 19:20:14,676] [INFO] [real_accelerator.py:161:get_accelerator] Setting ds_accelerator to cuda (auto detect)
--------------------------------------------------
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
 [WARNING]  async_io requires the dev libaio .so object and headers but these were not found.
 [WARNING]  async_io: please install the libaio-dev package with apt
 [WARNING]  If libaio is already installed (perhaps from source), try setting the CFLAGS and LDFLAGS environment variables to where it can be found.
async_io ............... [NO] ....... [NO]
fused_adam ............. [NO] ....... [OKAY]
cpu_adam ............... [NO] ....... [OKAY]
cpu_adagrad ............ [NO] ....... [OKAY]
cpu_lion ............... [NO] ....... [OKAY]
 [WARNING]  Please specify the CUTLASS repo directory as environment variable $CUTLASS_PATH
evoformer_attn ......... [NO] ....... [NO]
fused_lamb ............. [NO] ....... [OKAY]
fused_lion ............. [NO] ....... [OKAY]
inference_core_ops ..... [NO] ....... [OKAY]
cutlass_ops ............ [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
ragged_device_ops ...... [NO] ....... [OKAY]
ragged_ops ............. [NO] ....... [OKAY]
random_ltd ............. [NO] ....... [OKAY]
 [WARNING]  sparse_attn requires a torch version >= 1.5 and < 2.0 but detected 2.2
 [WARNING]  using untested triton version (2.2.0), only 1.0.0 is known to be compatible
sparse_attn ............ [NO] ....... [NO]
spatial_inference ...... [NO] ....... [OKAY]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
transformer_inference .. [NO] ....... [OKAY]
--------------------------------------------------
DeepSpeed general environment info:
torch install path ............... ['/opt/conda/envs/llama2env/lib/python3.8/site-packages/torch']
torch version .................... 2.2.2+cu121
deepspeed install path ........... ['/opt/conda/envs/llama2env/lib/python3.8/site-packages/deepspeed']
deepspeed info ................... 0.12.6, unknown, unknown
torch cuda version ............... 12.1
torch hip version ................ None
nvcc version ..................... 12.1
deepspeed wheel compiled w. ...... torch 2.1, cuda 12.1
shared memory (/dev/shm) size .... 25.55 GB
```

**System info (please complete the following information):**
 - OS: Linux Debian 10
 - GPU count and types: 2xT4s
 - (if applicable) Hugging Face Transformers/Accelerate/etc. versions: Please see above
 - Python version 3.8.15
 - Any other relevant info about your setup
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
Address sizes:       46 bits physical, 48 bits virtual
CPU(s):              8
On-line CPU(s) list: 0-7
Thread(s) per core:  2
Core(s) per socket:  4
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               79
Model name:          Intel(R) Xeon(R) CPU @ 2.20GHz
Stepping:            0
CPU MHz:             2199.998
BogoMIPS:            4399.99
Hypervisor vendor:   KVM
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            256K
L3 cache:            56320K
NUMA node0 CPU(s):   0-7
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities

**Docker context**
Are you using a specific docker image that you can share? 
No

**Comment by tohtana** [comment]
Hi @zcakzhuu, I ran the script a few times, but it produced exact same results. I used:
- deepspeed 0.16.2
- transformers 4.45.1
- flash-attn 2.7.2.post1

On the other hand, setting a seed is not sufficient to get stable results. See this [document](https://pytorch.org/docs/stable/notes/randomness.html) for details. You can enable the deterministic option by `transformers.enable_full_determinism` intead of `set_seed`.
